pallets/loans/src/types/cashflow.rs

```diff
@@ -0,0 +1,451 @@
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::{interest::InterestRate, Seconds};
+use chrono::{DateTime, Datelike, NaiveDate};
+use frame_support::pallet_prelude::RuntimeDebug;
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{
+		ensure_pow, EnsureAdd, EnsureAddAssign, EnsureFixedPointNumber, EnsureInto, EnsureSub,
+		EnsureSubAssign,
+	},
+	DispatchError, FixedPointNumber, FixedPointOperand, FixedU128,
+};
+use sp_std::{cmp::min, vec, vec::Vec};
+
+// By now only "day 1" of the month is supported for monthly cashflows.
+// Modifying this value would make `monthly_dates()` and
+// `monthly_intervals()` to no longer work as expected.
+// Supporting more reference dates will imply more logic related to dates.
+const DAY_1: u32 = 1;
+
+/// Specify the expected repayments date
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum Maturity {
+	/// Fixed point in time, in secs
+	Fixed {
+		/// Secs when maturity ends
+		date: Seconds,
+		/// Extension in secs, without special permissions
+		extension: Seconds,
+	},
+	/// No Maturity date
+	None,
+}
+
+impl Maturity {
+	pub fn fixed(date: Seconds) -> Self {
+		Self::Fixed { date, extension: 0 }
+	}
+
+	pub fn date(&self) -> Option<Seconds> {
+		match self {
+			Maturity::Fixed { date, .. } => Some(*date),
+			Maturity::None => None,
+		}
+	}
+
+	pub fn is_valid(&self, now: Seconds) -> bool {
+		match self {
+			Maturity::Fixed { date, .. } => *date > now,
+			Maturity::None => true,
+		}
+	}
+
+	pub fn extends(&mut self, value: Seconds) -> Result<(), DispatchError> {
+		match self {
+			Maturity::Fixed { date, extension } => {
+				date.ensure_add_assign(value)?;
+				extension.ensure_sub_assign(value)?;
+				Ok(())
+			}
+			Maturity::None => Err(DispatchError::Other(
+				"No maturity date that could be extended.",
+			)),
+		}
+	}
+}
+
+/// Interest payment periods
+#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum InterestPayments {
+	/// All interest is expected to be paid at the maturity date
+	None,
+
+	/// Interest is expected to be paid monthly
+	/// The associated value correspond to the paydown day in the month,
+	/// from 1-31.
+	/// The day will be adjusted to the month.
+	///
+	/// NOTE: Only day 1 is supported by now
+	Monthly(u8),
```

Comment:
I would remove that one, we are not using it atm, right?

Reply:
The variant is used. The value of the variant is clamped to expected to be `1`, but to avoid migrations, I'll keep it. WDYT?

Reply:
We will not use it and it complicates the logic IMO.

Reply:
If we add another variant we will not need a migration.

Reply:
But we will have a mixed logic.

```rust
InterestPayments {
    OnceAtmaturity,
    Monthly, // paid at day 1
    MonthlyAt(u8) // Added in the future, collides with MonthlyAt(1) == Monthly
}
```

I would prefer remove it, you know I like to remove things 😆, just being aware of the future migration

Reply:
I am not following. I would just remove all variants besides `OnceAtMaturity`. Then we can add it as you currently have it later on, if we want to have it the same way.

Reply:
But then we do not have monthly cashflows? At which moment do we have to change that requirement? That's the main point of this, isn't it?

Reply:
We decided that the current approach probably needs rethinking and hence only want to expose the runtime api plus the `OnceAtMaturity` payment. 
So that the UI and subquery can already index via the api and we have more time to make the feature right.

Reply:
I understood that we still wanted the monthly cashflows, but supposing all payments for all months were at maturity.

Reply:
> I understood that we still wanted the monthly cashflows, supposing all payments for all months were at maturity.

I do not think we will keep the `Monthly(u8)`. I would propose we move all test and existing code to a new branch and only bring in the runtim-api and the calcualtion for `OnceAtMaturity`.